javaslang/src/main/java/javaslang/collection/List.java

```diff
@@ -85,9 +81,80 @@
  * @author Daniel Dietrich
  * @since 1.1.0
  */
-public interface List<T> extends Kind1<List<?>, T>, LinearSeq<T>, Stack<T> {
+public final class List<T> implements Stack<T>, LinearSeq<T>, Kind1<List<?>, T>, Serializable {
```

Comment:
> Though I expected something else than "dude, you're not supposed to touch that, because that's how Haskell and Scala did it, so it's the way we should do it also - even if it might not be the best solution in Java".

Sorry! I was not prepared for such a change in this PR - it was a shocking moment and I was emotional. Changing the List from an ADT to a plain class has so much implications for the whole code base that many of my design decisions would be thrown over board.

There are different types of changes to the code base of Javaslang:

* reformatting - this shouldn't be done by committers in general (but solving inconsistencies is ok)
* adding features - this is the 'normal' case for PRs
* changing/optimizing implementations - also a 'normal' PR case
* performing design changes - this should be discussed in depth because it may have deep implications.

Design changes are very dangerous. Even if they are right they most probably will break backward compatibility. We need to be careful here and slow down a bit.

We should separate benchmarks and optimizations into multiple PRs instead of multiple commits. Is that possible without too much effort?